dpgen2/exploration/deviation/deviation_spin.py

```diff
@@ -0,0 +1,127 @@
+from collections import (
+    defaultdict,
+)
+from typing import (
+    Dict,
+    List,
+    Optional,
+)
+
+import numpy as np
+
+from .deviation_manager import (
+    DeviManager,
+)
+
+
+class DeviManagerSpin(DeviManager):
+    r"""The class which is responsible for DeepSPIN model deviation management.
+
+    This is the implementation of DeviManager for DeepSPIN model. Each deviation
+    (e.g. max_devi_af, max_devi_mf in file `model_devi.out`) is stored
+    as a List[Optional[np.ndarray]], where np.array is a one-dimensional
+    array.
+    A List[np.ndarray][ii][jj] is the force model deviation of the jj-th
+    frame of the ii-th trajectory.
+    The model deviation can be List[None], where len(List[None]) is
+    the number of trajectory files.
+
+    """
+
+    MAX_DEVI_AF = "max_devi_af"
+    MIN_DEVI_AF = "min_devi_af"
+    AVG_DEVI_AF = "avg_devi_af"
+    MAX_DEVI_MF = "max_devi_mf"
+    MIN_DEVI_MF = "min_devi_mf"
+    AVG_DEVI_MF = "avg_devi_mf"
+
+    def __init__(self):
+        super().__init__()
+        self._data = defaultdict(list)
+
+    def _check_name(self, name: str):
+        assert name in (
+            DeviManager.MAX_DEVI_V,
+            DeviManager.MIN_DEVI_V,
+            DeviManager.AVG_DEVI_V,
+            self.MAX_DEVI_AF,
+            self.MIN_DEVI_AF,
+            self.AVG_DEVI_AF,
+            self.MAX_DEVI_MF,
+            self.MIN_DEVI_MF,
+            self.AVG_DEVI_MF,
+        ), f"Error: unknown deviation name {name}"
+
+    def _add(self, name: str, deviation: np.ndarray) -> None:
+        assert isinstance(
+            deviation, np.ndarray
+        ), f"Error: deviation(type: {type(deviation)}) is not a np.ndarray"
+        assert len(deviation.shape) == 1, (
+            f"Error: deviation(shape: {deviation.shape}) is not a "
+            + f"one-dimensional array"
+        )
+        self._data[name].append(deviation)
+        self.ntraj = max(self.ntraj, len(self._data[name]))
+
+    def _get(self, name: str) -> List[Optional[np.ndarray]]:
+        if self.ntraj == 0:
+            return []
+        elif len(self._data[name]) == 0:
+            return [None for _ in range(self.ntraj)]
+        else:
+            return self._data[name]
+
+    def clear(self) -> None:
+        self.__init__()
+        return None
+
+    def _check_data(self) -> None:
+        r"""Check if data is valid"""
+        model_devi_names = (
+            DeviManager.MAX_DEVI_V,
+            DeviManager.MIN_DEVI_V,
+            DeviManager.AVG_DEVI_V,
+            self.MAX_DEVI_AF,
+            self.MIN_DEVI_AF,
+            self.AVG_DEVI_AF,
+            self.MAX_DEVI_MF,
+            self.MIN_DEVI_MF,
+            self.AVG_DEVI_MF,
```

Comment:
should be added to devi manager. 
what is the difference between `DeviManager.XXX_DEVI_F` and `self.XXX_DEVI_AF`? they should be referring to the same thing.